skyportal/plot.py

```diff
@@ -1398,397 +1400,423 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
     if len(spectra) == 0:
         return None, None, None
 
-    rainbow = cm.get_cmap('rainbow', len(spectra))
-    palette = list(map(rgb2hex, rainbow(range(len(spectra)))))
-    color_map = dict(zip([s.id for s in spectra], palette))
-
-    data = []
-    for i, s in enumerate(spectra):
-
-        # normalize spectra to a median flux of 1 for easy comparison
-        normfac = np.nanmedian(np.abs(s.fluxes))
-        normfac = normfac if normfac != 0.0 else 1e-20
-        altdata = json.dumps(s.altdata) if s.altdata is not None else ""
-        df = pd.DataFrame(
-            {
-                'wavelength': s.wavelengths,
-                'flux': s.fluxes / normfac,
-                'id': s.id,
-                'telescope': s.instrument.telescope.name,
-                'instrument': s.instrument.name,
-                'date_observed': s.observed_at.isoformat(sep=' ', timespec='seconds'),
-                'pi': (
-                    s.assignment.run.pi
-                    if s.assignment is not None
-                    else (
-                        s.followup_request.allocation.pi
-                        if s.followup_request is not None
-                        else ""
-                    )
-                ),
-                'origin': s.origin,
-                'altdata': altdata[:20] + "..." if len(altdata) > 20 else altdata,
-            }
+    layouts = []
+    spectrum_types = []
+    for spec_type in ALLOWED_SPECTRUM_TYPES:
+        spectra_of_same_type = [s for s in spectra if s.type == spec_type]
+        if len(spectra_of_same_type) == 0:
+            continue
+
+        rainbow = cm.get_cmap('rainbow', len(spectra_of_same_type))
+        palette = list(map(rgb2hex, rainbow(range(len(spectra_of_same_type)))))
+        color_map = dict(zip([s.id for s in spectra_of_same_type], palette))
+
+        spectrum_types.append(spec_type)
+
+        data = []
```

Comment:
The inner part of the loop would become easier to read if it is factored out into a separate function.